### PR TITLE
os/bluestore: allow dynamic levels

### PR DESCRIPTION
**Pull request by Huehueh**
os/bluestore: This PR introduces option bluestore_volume_selection_policy=without_slow_dir which helps avoiding OSD spillover and enables the user to turn on dynamic levels in RocksDB options.

More information about the PR is included in attached document:
[rocksdb_in_ceph.pdf](https://github.com/ceph/ceph/files/5224089/rocksdb_in_ceph.pdf)

Signed-off-by: Kinga Karczewska <kkarczewska@cloudferro.com>
Signed-off-by: Kajetan Janiak <kjaniak@cloudferro.com>

### CODE REVIEW

**Comment by xelexin** [comment]
@markhpc we talked about it, can you look at that?

**Comment by xelexin** [comment]
@ifed01 what do you think about this change? do you think is it a good way to fight against bluefs spillover?

**Comment by ifed01** [comment]
Generally LGTM as an experimental stuff for enabling dynamic RocksDB levels. Certainly we need to gain more expertise (primarily on dynamic levels) prior to making this a default behavior in production...
Just a nit - I don't like the name of the mode - may be something like "db volume first" instead?

**Review by ifed01 — APPROVED** [review]

**Review by ifed01 — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-09-16_14:28:48-rados-wip-kefu-testing-2020-09-16-1741-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-17_08:46:35-rados-wip-kefu-testing-2020-09-17-1122-distro-basic-smithi/

**Comment by tchaikov** [comment]
@ifed01 good to merge ? or you'd prefer waiting to see if @Huehueh wants to change the name of the mode?

**Comment by ifed01** [comment]
@Huehueh - what do you think about mode renaming?

**Comment by markhpc** [comment]
I'll try to look at this later today.  Adding @aclamk as I suspect his column family work that landed in octopus will already help in this area, though we can potentially still do better with dynamic levels.  I think @ifed01 is right though, this should be an experimental option and it's going to need lots of testing.

**Comment by Huehueh** [comment]
@ifed01 Hello, sorry for the delay. I will change the name. I just wanted to wait for opinions of others to see if there are any more comments concerning the code. I will make the commit today.

**Comment by Huehueh** [comment]
@ifed01 One more thing: I don't like "db volume first", because db is always first, we don't change that. What about "rocksdb_one_dir"?

**Comment by ifed01** [comment]
> @ifed01 One more thing: I don't like "db volume first", because db is always first, we don't change that. What about "rocksdb_one_dir"?


I'd prefer a name which refers to the high level meaning rather than one that covers implementation details. Hence all these references about "dirs" look doubtful to me as IMO they aren't obvious for users... May be something like "always use DB" or "Everything to DB" or "always try DB". 
Or maybe better ask native speakers... @markhpc - what do you think? Can you propose some other choices?

**Comment by Huehueh** [comment]
> > @ifed01 One more thing: I don't like "db volume first", because db is always first, we don't change that. What about "rocksdb_one_dir"?
> 
> I'd prefer a name which refers to the high level meaning rather than one that covers implementation details. Hence all these references about "dirs" look doubtful to me as IMO they aren't obvious for users... May be something like "always use DB" or "Everything to DB" or "always try DB".
> Or maybe better ask native speakers... @markhpc - what do you think? Can you propose some other choices?

I like 'always try db'. But let's wait for Marks opinion.

**Comment by aclamk** [comment]
> I like 'always try db'. But let's wait for Marks opinion.
After iterating few names of my own, I like this one.

**Review by aclamk — APPROVED** [review]
Nice optimization of DB volume.

**Comment by tchaikov** [comment]
@markhpc ping?

File: src/common/options.cc [hunk 1]
```diff
@@ -4687,9 +4687,12 @@ std::vector<Option> get_global_options() {
 
     Option("bluestore_volume_selection_policy", Option::TYPE_STR, Option::LEVEL_DEV)
     .set_default("use_some_extra")
-    .set_enum_allowed({ "rocksdb_original", "use_some_extra" })
+    .set_enum_allowed({ "rocksdb_original", "use_some_extra", "without_slow_dir" })
```
Comment (markhpc) [inline src/common/options.cc:4690]: My take here is that if we are trying to maximally scale levels to fit on the fast device, we should maybe use something like "fit_to_fast".  Not sure if that's clearer or not.

**Review by markhpc — APPROVED** [review]
All said this looks more or less fine to me, but the proof of whether or not this is a good idea will be in how well it actually works when using level_compaction_dynamic_level_bytes.  It would be nice to see some tests showcasing that it's worth adding the flexibility here.

File: src/common/options.cc (hunk 1 above)
Comment (Huehueh) [reply src/common/options.cc:4690]: I like this idea. "fit_to_fast" and "always_try_db" are both good options.

Comment (ifed01) [reply src/common/options.cc:4690]: let it be "fit_to_fast" then

**Comment by Huehueh** [comment]
> All said this looks more or less fine to me, but the proof of whether or not this is a good idea will be in how well it actually works when using level_compaction_dynamic_level_bytes. It would be nice to see some tests showcasing that it's worth adding the flexibility here.

@markhpc Do you have something specific in mind? Is there any specific pipeline? We have done some tests of this feature, it looked ok, but I agree we should do some more wide tests. Maybe we could use teuthology for that.

File: src/common/options.cc (hunk 1 above)
Comment (Huehueh) [reply src/common/options.cc]: I changed the name. It may appear that all the commits are new, but I only added my collegue to the signed-off.

**Comment by markhpc** [comment]
> > All said this looks more or less fine to me, but the proof of whether or not this is a good idea will be in how well it actually works when using level_compaction_dynamic_level_bytes. It would be nice to see some tests showcasing that it's worth adding the flexibility here.
> 
> @markhpc Do you have something specific in mind? Is there any specific pipeline? We have done some tests of this feature, it looked ok, but I agree we should do some wider tests. Maybe we could use teuthology for that.

Off the top of my head: 

rados bench can generate a moderate amount of metadata in rocksdb (feeding it longer object names probably should make it grow faster) and is fairly easy to run.  You could do rbd+fio but that requires a lot more data.  RGW generates even more with bucket indexes but iit's a little more work to setup.  It could be interesting to compare what happens with column family sharding off, column family sharding on, and column family sharding with your PR and some reasonable level_compaction_dynamic_level_bytes setup.  Just vstart or cbt with a single osd would be fine.  Goal would be to look at performance, compaction statistics from the osd log via https://github.com/ceph/cbt/blob/master/tools/ceph_rocksdb_log_parser.py and perhaps utilization of the fast/slow devices by level.

**Comment by Huehueh** [comment]
> > > All said this looks more or less fine to me, but the proof of whether or not this is a good idea will be in how well it actually works when using level_compaction_dynamic_level_bytes. It would be nice to see some tests showcasing that it's worth adding the flexibility here.
> > 
> > 
> > @markhpc Do you have something specific in mind? Is there any specific pipeline? We have done some tests of this feature, it looked ok, but I agree we should do some wider tests. Maybe we could use teuthology for that.
> 
> Off the top of my head:
> 
> rados bench can generate a moderate amount of metadata in rocksdb (feeding it longer object names probably should make it grow faster) and is fairly easy to run. You could do rbd+fio but that requires a lot more data. RGW generates even more with bucket indexes but iit's a little more work to setup. It could be interesting to compare what happens with column family sharding off, column family sharding on, and column family sharding with your PR and some reasonable level_compaction_dynamic_level_bytes setup. Just vstart or cbt with a single osd would be fine. Goal would be to look at performance, compaction statistics from the osd log via https://github.com/ceph/cbt/blob/master/tools/ceph_rocksdb_log_parser.py and perhaps utilization of the fast/slow devices by level.

We had some tests using benchmarks with vstart on simple rgw and it turned out quite well. I will create some summary about it and perhaps conduct some more tests.

**Comment by markhpc** [comment]
@Huehueh Sounds good!

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-17_08:59:57-rados-wip-kefu-testing-2020-10-17-1516-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-18_16:30:00-rados-wip-kefu-testing-2020-10-18-2239-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-19_04:31:12-rados-wip-kefu-testing-2020-10-18-2239-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-19_06:06:20-rados-wip-kefu-testing-2020-10-18-2239-distro-basic-smithi

failures tracked by

- https://tracker.ceph.com/issues/47024

**Comment by Huehueh** [comment]
Sorry that it takes so long, but during tests of our feature we have discovered a thing that I suppose can be a bug. We made some tests with `bluestore_rocksdb_cf` set to `true`and we noticed that db usage taken using `ceph daemon osd.num perf dump` is exactly two times bigger (8GB) than sum of data on all RocksDb levels taken from Compaction Stats (please see image 1). I suppose it is a bug because my test included benchmark of about 4GB data and if I read allocations from log file it is also around 4GB (see second image). This causes early spillover also in some cases. This situation only happens if there are multiple column families, so please @aclamk could you tell me if it is desired behaviour? 
image2
image2

**Comment by Huehueh** [comment]
@aclamk Could you please respond?

**Comment by aclamk** [comment]
@Huehueh 
I think I will try to replay your steps.
1) Are you changing bluestore_rocksdb_options from default: "compression=kNoCompression,max_write_buffer_number=4,min_write_buffer_number_to_merge=1,recycle_log_file_num=4,write_buffer_size=268435456,writable_file_max_buffer_size=0,compaction_readahead_size=2097152,max_background_compactions=2,max_total_wal_size=1073741824" ?
2) can you list files on BlueFS when usage is high? "ceph daemon osd.num bluefs files list"
Is there a lot of .wal files or .sst files?

**Comment by Huehueh** [comment]
> @Huehueh
> I think I will try to replay your steps.
> 
>     1. Are you changing bluestore_rocksdb_options from default: "compression=kNoCompression,max_write_buffer_number=4,min_write_buffer_number_to_merge=1,recycle_log_file_num=4,write_buffer_size=268435456,writable_file_max_buffer_size=0,compaction_readahead_size=2097152,max_background_compactions=2,max_total_wal_size=1073741824" ?
> 
>     2. can you list files on BlueFS when usage is high? "ceph daemon osd.num bluefs files list"
>        Is there a lot of .wal files or .sst files?

Hi, thanks for response. I created a ticket for that where I described the problem more carefully, please look at it: https://tracker.ceph.com/issues/48070

As to your questions:
1. My RocksDB settings were:
'num_levels=5,max_bytes_for_level_base=268435456,max_bytes_for_level_multiplier=10,level0_file_num_compaction_trigger=6,max_background_jobs=5,max_compaction_bytes=104857600,write_buffer_size=22369621,max_write_buffer_number=5,min_write_buffer_number_to_merge=2,compression=kNoCompression,compaction_style=kCompactionStyleLevel,stats_dump_period_sec=1'
and as I told before DB size was: bluestore block db size = 10737418240

2. I will recreate the test to see it and I will respond tomorrow.

**Comment by Huehueh** [comment]
@aclamk Ok, I recreated the test. The results are quite similar (see attached image). 
There are 79 sst and 158 wal files. I attach the output of the command  "ceph daemon osd.num bluefs files list" (file_list.txt).
I attached also the Ceph options I set during the test, so you could recreate it (file options.txt)

image
[file_list.txt](https://github.com/ceph/ceph/files/5496737/file_list.txt)
[options.txt](https://github.com/ceph/ceph/files/5497222/options.txt)

**Comment by Huehueh** [comment]
@aclamk Ok, I recreated the test. The results are quite similar (see attached image).
There are 79 sst and 158 wal files. I attach the output of the command "ceph daemon osd.num bluefs files list" (file_list.txt).
I attached also the Ceph options I set during the test, so you could recreate it (file options.txt)
image
[file_list.txt](https://github.com/ceph/ceph/files/5497230/file_list.txt)
[options.txt](https://github.com/ceph/ceph/files/5497244/options.txt)

Sorry, I closed the PR by accident:) I reopened it.

**Comment by Huehueh** [comment]
@aclamk Any update on the subject?

**Comment by aclamk** [comment]
@Huehueh 
1) I have tested:
- this PR default
- this PR with fit_to_fast selected
Multiplied by:
- sharding on
- sharding off

I DO observe fluctuations on DB used and WAL used but within reasonable sizes.
https://docs.google.com/spreadsheets/d/1bPna9MCrIkSL7kJ8hArxIujBp3_9I9V_ljRVTsnCF6s/edit?usp=sharing

2) In your configuration `bluestore_rocksdb_options` I do not see `max_total_wal_size` set, which is present in default.
https://github.com/ceph/ceph/pull/35277

3) How large is your WAL space? Do I read correctly from bluefs files that it is ~100MB ?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -5601,33 +5601,39 @@ int BlueStore::_open_bluefs(bool create, bool read_only)
...
       rocks_opts,
       nullptr);
     if (r < 0) {
       return r;
     }
-
-    double reserved_factor = cct->_conf->bluestore_volume_selection_reserved_factor;
-    vselector =
-      new RocksDBBlueFSVolumeSelector(
+    if (cct->_conf->bluestore_volume_selection_policy == "without_slow_dir") {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5618]: This must be synchronized with available options.

**Review by aclamk — CHANGES_REQUESTED** [review]
Must synchronize strings in-code with options.

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (Huehueh) [reply src/os/bluestore/BlueStore.cc]: Very true, thank you for pointing this out. I don't know how I missed it. I corrected it just now.

**Comment by Huehueh** [comment]
@aclamk Hello, thank you for your responses. It it true that we had 100MB for WAL and it turned out when I set 1GB for  `max_total_wal_size` in RocksDB options and 1GB for `bluestore block wal size`, the results were ok - DB usage was more or less the same as the usage shown by compaction stats. Than you for your help. I did not expect that WAL is the problem here. Nevertheless I think that the size of WAL should not have influence on DB data, that is why I will keep the issue https://tracker.ceph.com/issues/48070 open.

**Comment by Huehueh** [comment]
@markhpc @ifed01 @aclamk After some help from Adam I could perform my tests and show some reasonable results. 

**Test 1**: Write benchmarks (omaps) of about 4GB of data in 5MB chunks.
Ceph settings: DB 10GB, WAL 1 GB, L1=default (256MB), no compression, `write_buffer_size` ~20MB, dynamic levels OFF when `fit-to-fast`.
Results: 
[bluestore_volume_selection_policy - test 1.pdf](https://github.com/ceph/ceph/files/5566619/bluestore_volume_selection_policy.-.test.1.pdf)
Summary: 
- bluestore_rocksdb_cf = false: When `rocksdb_original` is selected, there is spillover when DB is filled in about 25% (data enters L3). We see that `fit_to_fast` performance is comparable to `use_some_extra`, there is no spillover. 
- bluestore_rocksdb_cf = false: There is no spillover because data is distributed (unevenly) amid multiple column families. 

**Test 2** Write benchmarks of about 3GB of omap data in 5MB chunks, then 50 times write and delete 2 omaps of 5MB alternately.
Ceph settings: DB 4GB, WAL 1GB, `bluestore_rocksdb_cf=true`, L1 default (256MB), `write_buffer_size` ~20MB, `level_compaction_dynamic_level_bytes=true` when `fit-to-fast`.
Results: 
[bluestore_volume_selection_policy - test 2.pdf](https://github.com/ceph/ceph/files/5566621/bluestore_volume_selection_policy.-.test.2.pdf)
Summary:
- When `rocksdb_original` or `use_some_extra` options are set, the spillover occurs quite early. In these cases many column families have the same RocksDB levels sizes although they are filled to varying degrees. 
- In case when `fit_to_fast` is on, spillover is moved away in time because levels sizes are better adapted to current column family filling. This could happen only when `level_compaction_dynamic_level_bytes` is set to true (only possible with `fit_to_fast`).

**Comment by ifed01** [comment]
> @markhpc @ifed01 @aclamk After some help from Adam I could perform my tests and show some reasonable results.
> 
> **Test 2** Write benchmarks of about 3GB of omap data in 5MB chunks, then 50 times write and delete 2 omaps of 5MB alternately.
> Ceph settings: DB 4GB, WAL 1GB, `bluestore_rocksdb_cf=true`, L1 default (256MB), `write_buffer_size` ~20MB, `level_compaction_dynamic_level_bytes=true` when `fit-to-fast`.
> Results:
> [bluestore_volume_selection_policy - test 2.pdf](https://github.com/ceph/ceph/files/5566621/bluestore_volume_selection_policy.-.test.2.pdf)
> Summary:
> 
> * When `rocksdb_original` or `use_some_extra` options are set, the spillover occurs quite early. In these cases many column families have the same RocksDB levels sizes although they are filled to varying degrees.
> * In case when `fit_to_fast` is on, spillover is moved away in time because levels sizes are better adapted to current column family filling. This could happen only when `level_compaction_dynamic_level_bytes` is set to true (only possible with `fit_to_fast`).

Wondering if you adjusted "bluestore_volume_selection_reserved_factor" in this experiment. By default it's set to 2x which means that DB volume needs to be larger than 5GB for "use_some_extra" mode to be in action. With 4GB DB volume there is no "extra" space to use....

**Comment by Huehueh** [comment]
> > @markhpc @ifed01 @aclamk After some help from Adam I could perform my tests and show some reasonable results.
> > **Test 2** Write benchmarks of about 3GB of omap data in 5MB chunks, then 50 times write and delete 2 omaps of 5MB alternately.
> > Ceph settings: DB 4GB, WAL 1GB, `bluestore_rocksdb_cf=true`, L1 default (256MB), `write_buffer_size` ~20MB, `level_compaction_dynamic_level_bytes=true` when `fit-to-fast`.
> > Results:
> > [bluestore_volume_selection_policy - test 2.pdf](https://github.com/ceph/ceph/files/5566621/bluestore_volume_selection_policy.-.test.2.pdf)
> > Summary:
> > 
> > * When `rocksdb_original` or `use_some_extra` options are set, the spillover occurs quite early. In these cases many column families have the same RocksDB levels sizes although they are filled to varying degrees.
> > * In case when `fit_to_fast` is on, spillover is moved away in time because levels sizes are better adapted to current column family filling. This could happen only when `level_compaction_dynamic_level_bytes` is set to true (only possible with `fit_to_fast`).
> 
> Wondering if you adjusted "bluestore_volume_selection_reserved_factor" in this experiment. By default it's set to 2x which means that DB volume needs to be larger than 5GB for "use_some_extra" mode to be in action. With 4GB DB volume there is no "extra" space to use....

@ifed01 That's right, I forgot that this feature requires this. I assume that we need at least 7 GB (L0 +L1 + 2 * L2 = 1GB + 256MB +2*2,56GB - as it assumes L0 has 1GB Bluestore.cc l. 5657 on master) so I performed test 3 with DB = 7GB.
Although I must say that this calculation is rather adapted to one column family, rather than 8, right?

**Test 3** Write benchmarks of about 5,6 GB of omap data in 5MB chunks, then100 times write 4 omaps and delete 4 random omaps of 5MB alternately.
Ceph settings: DB 7GB, WAL 1GB, bluestore_rocksdb_cf=true, L1 default (256MB), write_buffer_size ~20MB, level_compaction_dynamic_level_bytes=true when fit-to-fast.
Results:
[bluestore_volume_selection_policy - test 3.pdf](https://github.com/ceph/ceph/files/5582396/bluestore_volume_selection_policy.-.test.3.pdf)

Summary:
- When `rocksdb_original` is set as the `bluestore_volume_selection_policy` the final spillover is about 3 GB
- When `use_some_extra` is set - spillover is about 2,2 GB.
- When `fit_to_fast` is set - spillover is about 1,9 GB.
Summary: Results from `use_some_extra` and  `fit_to_fast` are similar, although the proposed feature `fit_to_fast` is simpler (the user does not have to adapt "reserved" factor) and it is better suited for multiple column families (it does not assume any size of levels in calculations). 

To sum up, the tests showed that the proposed feature has good impact on RocksDB usage and can be proposed for the user as an alternative. 
@aclamk Re-review the changes I made, please.

**Review by aclamk — APPROVED** [review]
This is good addition to bluefs volume selection options.

**Comment by Huehueh** [comment]
Thank you very much! I added signed-off to my last commit, so I hope now PR will be merged!

**Comment by Huehueh** [comment]
@tchaikov Could you please merge this PR?

**Review by tchaikov — CHANGES_REQUESTED** [review]
@Huehueh this PR has been approved.  but it deserves a rados qa run before being merged.

also, i found that the last commit ("changed all occurances of without_slow_dir to fit_to_fast") fixes one of the previous commits in this very PR. could you please squash it into the commit introducing "fit_to_fast"?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-04_09:24:21-rados-wip-kefu-testing-2020-12-04-1516-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-05_08:42:04-rados-wip-kefu-testing-2020-12-05-1217-distro-basic-smithi/

**Comment by Huehueh** [comment]
> @Huehueh this PR has been approved. but it deserves a rados qa run before being merged.
> 
> also, i found that the last commit ("changed all occurances of without_slow_dir to fit_to_fast") fixes one of the previous commits in this very PR. could you please squash it into the commit introducing "fit_to_fast"?

Ok, I will squash it. Thank you!

File: src/os/bluestore/BlueFS.cc [hunk 3]
```diff
@@ -3594,3 +3594,10 @@ void OriginalVolumeSelector::dump(ostream& sout) {
     << ", slow_total:" << slow_total
     << std::endl;
 }
+
+// ===============================================
+// WithoutSlowDirVolumeSelector
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc:3599]: @Huehueh - mind rename this class to something like FitToFastVolumeSelector?

Comment (Huehueh) [reply src/os/bluestore/BlueFS.cc]: Ok, no problem!

**Comment by Huehueh** [comment]
@tchaikov Could you please review and accept the changes?
